### ISSUE TITLE
reshuffle the mini-batches per epoch

### ISSUE THREAD

**Issue by brucefan1983** [brucefan1983/GPUMD#241]
Currently, we only shuffle the structures in the training data set at the beginning of the training. Perhaps it is better to shuffle for each epoch.

**Comment by brucefan1983** [comment]
The figure below was obtained by testing a very abundent silicon data set (from a GAP paper). The conclusions are:
* Full-batch gives the best accuracy if we can train up to the final convergence.
* Shuffling frequently does not improve the training results. It actaully do the opposite.
* Shuffling in the beginning only works well and the results with batchsize=100 is already quite close to those with full-batch.

![image](https://user-images.githubusercontent.com/24891193/189520873-f5c00ae5-5335-49d1-b2f8-8a8d360e4d84.png)

**Comment by brucefan1983** [comment]
If I change the x-axis from the number of generations to the computation time, the figure looks like as follows. It can be seen that using a mini-batch can effectively reduce the training time. Therefore, we suggest:
* use a proper mini-batch with a batch size from 100 to 1000 during the testing stage, when there is still potential to add more structures to the training data set.
* use the full-batch for the final training after the data set is completely fixed.
![image](https://user-images.githubusercontent.com/24891193/189521104-6a15632e-9a98-4296-a2ff-ea59d6861f99.png)

**Comment by brucefan1983** [comment]
In summary, using  a mini-batch migth lose 10% accuracy as compared to the full-batch training, but can reduce the training time by many times if the training data set is large (or huge).
